cpp/src/parquet/arrow/schema.cc

```diff
@@ -839,7 +839,8 @@ std::function<std::shared_ptr<::arrow::DataType>(FieldVector)> GetNestedFactory(
   switch (inferred_type.id()) {
     case ::arrow::Type::STRUCT:
       if (origin_type.id() == ::arrow::Type::STRUCT) {
-        return ::arrow::struct_;
+        return static_cast<std::shared_ptr<::arrow::DataType> (*)(const FieldVector&)>(
+            &::arrow::struct_);
```

Comment:
Or can use a lambda for clarity:
```suggestion
        return [](FieldVector fields) {
          return ::arrow::struct_(std::move(fields));
        };
```